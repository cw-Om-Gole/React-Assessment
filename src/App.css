* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  font-family: Arial, sans-serif;
}

.container {
  display: grid;
  grid-template-columns: 1fr 3fr;
  gap: 20px;
  padding: 20px;
  margin: 50px auto;
  max-width: 1024px;
}

.filters {
  background-color: white;
  padding: 20px;
  border: 1px solid #e0e0e0;
  border-radius: 5px;
  position: sticky;
  top: 20px; /* Stick to the top of the viewport */
  height: 60vh;
}

.filters h2 {
  font-size: 18px;
  font-weight: bold;
  margin-bottom: 10px;
}

.clear-all {
  display: block;
  text-align: right;
  color: #007bff;
  font-size: 14px;
  margin-bottom: 20px;
  cursor: pointer;
}

.filter-group {
  margin-bottom: 20px;
}

.filter-group label {
  display: block;
  font-size: 14px;
  margin-bottom: 5px;
}
 
.filter-group select {
  width: 100%;
  padding: 8px;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.filter-group input[type="checkbox"] {
  margin-right: 5px;
}

.main-content {
  background-color: white;
  padding: 20px;
  border: 1px solid #e0e0e0;
  border-radius: 5px;
}

.filter-container {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.filter-container a {
  text-decoration: none;
}



.budget-container {
  display: flex;
  justify-content: space-between;
}

.budget-container input {
  width: 40%;
  padding: 10px;
}

.budget-container input:active {
  border: 1px rgb(50, 180, 231) solid;
}

.budget-container span {
  align-self: center;
}




.sort {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  margin-bottom: 20px;
  border: 1px solid #ddd;
  padding: 10px;
  border-radius: 5px;
}


.sort .label {
  font-size: 16px;
  margin-right: 10px;
}

.sort select {
  padding: 8px;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.products-grid {
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 20px;
}

.product-card {
  background-color: #f9f9f9;
  padding: 20px;
  border: 1px solid #e0e0e0;
  border-radius: 5px;
  text-align: center;
}

.product-card img {
  width: 100%;
  height: 150px;
  object-fit: cover;
  margin-bottom: 15px;
}

.product-card h3 {
  font-size: 16px;
  font-weight: bold;
  margin-bottom: 10px;
}

.product-card p {
  font-size: 14px;
  color: #777;
  margin-bottom: 10px;
}

.product-card .price {
  font-size: 18px;
  font-weight: bold;
  color: #333;
  margin-bottom: 15px;
}

.product-card button {
  padding: 10px;
  width: 100%;
  background-color: #ff4d4d;
  color: white;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  transition: background-color 0.3s;
}

.product-card button:hover {
  background-color: #e60000;
}
