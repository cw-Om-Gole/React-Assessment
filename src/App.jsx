import { useEffect, useState } from 'react'
import './App.css'
import Card from './components/card'
import Fueltype from './components/Fueltype';
import axios from 'axios'

const fuelMap = {
  petrol: 1,
  diesel: 2,
  cng: 3,
  lpg: 4,
  electric: 5,
  hybrid: 6
}

const baseURL = 'api/stocks';

function App() {
  const [cars, setCars] = useState([]);
  const [loading, setLoading] = useState(true);
  const [apiURL, setApiURL] = useState(baseURL);
  const [sortOrder, setSortOrder] = useState('default');
  const [selectedFuels, setSelectedFuels] = useState([]);
  const [min, setMin] = useState("");
  const [max, setMax] = useState("");

  const handleSortChange = (e) => {
    setSortOrder(e.target.value);
  }

  const handleMax = (e) => {
    setMax(e.target.value);
  }

  const handleMin = (e) => {
    setMin(e.target.value);
  }

  const getCars = async () => {
    try {
      const response = await fetch(apiURL);
      const carData = await response.json();

      setCars(carData.stocks);
      setLoading(false);
    } catch (error) {
      console.log(error);
    }

  }

  useEffect(() => {
    getCars();
    // if (sortOrder !== "default") handleSort(cars);
  }, [apiURL]);

  useEffect(() => {
    let newFuelString = "";

    if (selectedFuels.length !== 0) {
      newFuelString = `fuel=${selectedFuels.join('+')}`;
    }

    let newBudgetString = "";



    if (max !== "") {
      if (min === "") {
        newBudgetString = `budget=0-${max}`;
      } else {
        if (Number(min) > Number(max)) {
          alert("Invalid Input");
          return;
        }
        newBudgetString = `budget=${min}-${max}`;
      }
    }

    if (min !== "" && max === "") {
      newBudgetString = `budget=${min}-100000000`;
    }

    if (newBudgetString === "" && newFuelString === "") return;

    if (newBudgetString === "") {
      setApiURL(baseURL + "?" + newFuelString);
    } else if (newFuelString === "") {
      setApiURL(baseURL + "?" + newBudgetString);
    } else {
      setApiURL(baseURL + "?" + newFuelString + "&" + newBudgetString);
    }

  }, [selectedFuels, min, max]);

  // const handleSort = (cars) => {
  //   if(sortOrder === "default") {
  //     getCars();
  //     return;
  //   }

  //   console.log("sort");

  //   const sortedCars = [...cars].sort((a, b) => {
  //     if (sortOrder === 'asc') {
  //       return Number(a.priceNumeric) - Number(b.priceNumeric);
  //     } else if (sortOrder === 'desc') {
  //       return Number(b.priceNumeric) - Number(a.priceNumeric);
  //     }

  //     return 0;
  //   });

  //   setCars(sortedCars);
  // }

  const handleClear = (e) => {
    e.preventDefault();
    setApiURL(baseURL);
    setMin("");
    setMax("");
    setSelectedFuels([]);
    setSortOrder("default");
  }

  // useEffect(() => {
  //   handleSort(cars);
  // }, [sortOrder]);

  useEffect(() => {
    console.log(cars);
  }, [cars])

  if (loading) {
    return <div>Loading</div>;
  }

  if (cars.length == 0) {
    return <div>No cars found</div>;
  }

  return (
    <div className="container">
      <div className="filters">
        <div className="filter-container">
          <h2>Filters</h2>
          <a className="clear-all" onClick={handleClear}>Clear All</a>
        </div>

        <div className="filter-group">
          <label>Budget (Lakh)</label>
          <div className="budget-container">
            <input type="number" min="0" value={min} onChange={handleMin} />
            <span> - </span>
            <input type="number" min="1" value={max} onChange={handleMax} />
          </div>
        </div>


        <div className="filter-group">
          <h2>Fuel</h2>

          {Object.entries(fuelMap).map(([key, value]) => (
            <Fueltype fueltype={key} fuelNumber={value} selectedFuels={selectedFuels} setSelectedFuels={setSelectedFuels} />
          ))}

        </div>

      </div>

      <div className="main-content">
        <div className="sort">
          <div className="label">Sort By:</div>
          <select value={sortOrder} onChange={handleSortChange}>
            <option value="default">Default</option>
            <option value="asc">Price - Low to High</option>
            <option value="desc">Price - High to Low</option>
          </select>
        </div>

        <div className="products-grid">
          {cars.length != 0 && cars.sort((a, b) => {
            if (sortOrder === 'asc') {
              return Number(a.priceNumeric) - Number(b.priceNumeric);
            } else if (sortOrder === 'desc') {
              return Number(b.priceNumeric) - Number(a.priceNumeric);
            }

            return 0;
          }).map(car => <Card carDetails={car} />)}

        </div>
      </div>
    </div>)
}

export default App;
