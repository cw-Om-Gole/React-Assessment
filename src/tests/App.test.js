import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import App from '../App';


describe('App Component 1', () => {
    test('matches snapshot', async() => {
        const { asFragment } = render(<App />);
        
        expect(asFragment()).toMatchSnapshot();
    }); 

    test('filters cars by budget', async () => {
        render(<App />);
      
        const minInput = screen.getByPlaceholderText('0');
        const maxInput = screen.getByPlaceholderText('21');

        fireEvent.change(minInput, { target: { value: '10000' } });
        fireEvent.change(maxInput, { target: { value: '20000' } });

        expect(minInput.value).toBe('10000');
        expect(maxInput.value).toBe('20000');
    });

    test('sort cars by budget', async () => {
        render(<App />);
      
        const minInput = screen.getByPlaceholderText('0');
        const maxInput = screen.getByPlaceholderText('21');

        fireEvent.change(minInput, { target: { value: '10000' } });
        fireEvent.change(maxInput, { target: { value: '20000' } });

        expect(minInput.value).toBe('10000');
        expect(maxInput.value).toBe('20000');
    });

    test('changes sort order to "Price - Low to High"', async () => {
        render(<App />);
        await waitFor(() => {
          const sortSelect = screen.getByRole('combobox');
          fireEvent.change(sortSelect, { target: { value: 'asc' } });
    
          expect(sortSelect).toHaveValue('asc');
        });
      });
    
      test('changes sort order to "Price - High to Low"', async () => {
        render(<App />);
        await waitFor(() => {
          const sortSelect = screen.getByRole('combobox');
          fireEvent.change(sortSelect, { target: { value: 'desc' } });
    
          expect(sortSelect).toHaveValue('desc');
        });
      });

      test('changes sort order to "Default"', async () => {
        render(<App />);
        await waitFor(() => {
          const sortSelect = screen.getByRole('combobox');
          fireEvent.change(sortSelect, { target: { value: 'default' } });
    
          expect(sortSelect).toHaveValue('default');
        });
      });
});


