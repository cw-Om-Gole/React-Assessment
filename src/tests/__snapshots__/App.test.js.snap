// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component 1 matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="filters"
    >
      <div
        class="filter-container"
      >
        <h2>
          Filters
        </h2>
        <a
          class="clear-all"
        >
          Clear All
        </a>
      </div>
      <div
        class="filter-group"
      >
        <label
          for="budget"
        >
          Budget
        </label>
        <div
          class="budget-container"
        >
          <input
            id="budget"
            min="0"
            placeholder="0"
            type="number"
            value=""
          />
          <span>
             - 
          </span>
          <input
            id="budget"
            min="1"
            placeholder="21"
            type="number"
            value=""
          />
        </div>
      </div>
      <div
        class="filter-group"
      >
        <h2>
          Fuel
        </h2>
        <div
          class="checkbox-container"
        >
          <input
            id="petrol"
            name="petrol"
            type="checkbox"
          />
          <label
            for="petrol"
          >
            Petrol
          </label>
        </div>
        <div
          class="checkbox-container"
        >
          <input
            id="diesel"
            name="diesel"
            type="checkbox"
          />
          <label
            for="diesel"
          >
            Diesel
          </label>
        </div>
        <div
          class="checkbox-container"
        >
          <input
            id="cng"
            name="cng"
            type="checkbox"
          />
          <label
            for="cng"
          >
            Cng
          </label>
        </div>
        <div
          class="checkbox-container"
        >
          <input
            id="lpg"
            name="lpg"
            type="checkbox"
          />
          <label
            for="lpg"
          >
            Lpg
          </label>
        </div>
        <div
          class="checkbox-container"
        >
          <input
            id="electric"
            name="electric"
            type="checkbox"
          />
          <label
            for="electric"
          >
            Electric
          </label>
        </div>
        <div
          class="checkbox-container"
        >
          <input
            id="hybrid"
            name="hybrid"
            type="checkbox"
          />
          <label
            for="hybrid"
          >
            Hybrid
          </label>
        </div>
      </div>
    </div>
    <div
      class="main-content"
    >
      <div
        class="sort"
      >
        <div
          class="label"
        >
          Sort By:
        </div>
        <select>
          <option
            value="default"
          >
            Default
          </option>
          <option
            value="asc"
          >
            Price - Low to High
          </option>
          <option
            value="desc"
          >
            Price - High to Low
          </option>
        </select>
      </div>
      <div
        class="products-grid"
      />
    </div>
  </div>
</DocumentFragment>
`;
